### PR TITLE
Add a no commit to specific branch hook.

### PR DESCRIPTION
**Pull request by karcaw**

### CODE REVIEW

File: pre_commit_hooks/no_commit_to_branch.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+from pre_commit_hooks.util import cmd_output
+
+
+def is_on_branch(protected):
+    retval = False
+    branch = cmd_output('git', 'symbolic-ref', 'HEAD')
+    chunks = branch.strip().split('/')
+    if chunks[2] == protected:
+        retval = True
+    return retval
```
Comment (asottile) [inline pre_commit_hooks/no_commit_to_branch.py]: this `if ...: return True return False` can just be `return chunks[2] == protected`

File: pre_commit_hooks/no_commit_to_branch.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def main(argv=None):
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-b', default='master', help='branch to disallow commits to')
+    parser.add_argument('filenames', nargs='*', help='filenames to check.')
```
Comment (asottile) [inline pre_commit_hooks/no_commit_to_branch.py]: note for `always_run` hooks, they will receive no arguments

File: pre_commit_hooks/no_commit_to_branch.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    args = parser.parse_args(argv)
+
+    if is_on_branch(args.b):
+        return 1
+    else:
+        return 0
```
Comment (asottile) [inline pre_commit_hooks/no_commit_to_branch.py]: could be simplified to `return int(is_on_branch(args.b))`

File: pre_commit_hooks/no_commit_to_branch.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    chunks = branch.strip().split('/')
+    if chunks[2] == protected:
+        retval = True
+    return retval
+
+
+def main(argv=None):
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-b', default='master', help='branch to disallow commits to')
```
Comment (asottile) [inline pre_commit_hooks/no_commit_to_branch.py]: maybe add a longopt `--branch` so the code can read easier `args.branch`

File: setup.py [hunk 5]
```diff
@@ -55,6 +55,7 @@
             'pretty-format-json = pre_commit_hooks.pretty_format_json:pretty_format_json',
             'requirements-txt-fixer = pre_commit_hooks.requirements_txt_fixer:fix_requirements_txt',
             'trailing-whitespace-fixer = pre_commit_hooks.trailing_whitespace_fixer:fix_trailing_whitespace',
+            'no-commit-to-branch = pre_commit_hooks.no_commit_to_branch:main',
```
Comment (asottile) [inline setup.py]: can you alphabetize this?

File: pre_commit_hooks/no_commit_to_branch.py (hunk 3 above)
Comment (karcaw) [inline pre_commit_hooks/no_commit_to_branch.py]: I like that better too

File: pre_commit_hooks/no_commit_to_branch.py (hunk 2 above)
Comment (karcaw) [inline pre_commit_hooks/no_commit_to_branch.py]: removing..

File: pre_commit_hooks/no_commit_to_branch.py (hunk 4 above)
Comment (karcaw) [inline pre_commit_hooks/no_commit_to_branch.py]: ok!

File: pre_commit_hooks/no_commit_to_branch.py (hunk 1 above)
Comment (karcaw) [inline pre_commit_hooks/no_commit_to_branch.py]: removes a few lines of code too.

File: setup.py (hunk 5 above)
Comment (karcaw) [inline setup.py]: yes

File: .pre-commit-hooks.yaml [hunk 6]
```diff
@@ -147,3 +147,9 @@
     entry: trailing-whitespace-fixer
     language: python
     files: \.(asciidoc|adoc|coffee|cpp|css|c|ejs|erb|groovy|h|haml|hh|hpp|hxx|html|in|j2|jade|json|js|less|markdown|md|ml|mli|pp|py|rb|rs|R|scala|scss|sh|slim|tex|tmpl|ts|txt|yaml|yml)$
+-   id: no-commit-to-branch
+    name: Dont commit to branch
```
Comment (chriskuehl) [inline .pre-commit-hooks.yaml]: nit: apostrophe in "Don't"? You might have to fool YAML by not using a bareword, e.g.

```yaml
name: "Don't commit to branch"
```

Comment (karcaw) [inline .pre-commit-hooks.yaml]: that seemed to work, I went ahead and put the hook in alphabetic order too.

File: pre_commit_hooks/no_commit_to_branch.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import argparse
+import sys
+
+from pre_commit_hooks.util import cmd_output
+
+
+def is_on_branch(protected):
+    branch = cmd_output('git', 'symbolic-ref', 'HEAD')
+    chunks = branch.strip().split('/')
+    return chunks[2] == protected
```
Comment (asottile) [inline pre_commit_hooks/no_commit_to_branch.py]: oops sorry I didn't catch this at first, I think this should be:

```python
return '/'.join(chunks[2:]) == protected
```

branches can have slashes in them (maybe add a test for this too?)

Comment (karcaw) [inline pre_commit_hooks/no_commit_to_branch.py]: Adding Tests for this change.

**Comment by asottile** [comment]
Thanks for the PR!

I'm currently fixing the windows tests, they're grumpy about the newly added non-utf8 file

**Pull request merged**